### ISSUE TITLE
Implement auto continue as new 

### ISSUE THREAD

**Issue by longquanzheng** [indeedeng/iwf#107]
~~phase 1: conditional auto~~ 
* ~~only trigger when threshold meets and all pending states not started, inflight states= 0~~
* ~~single thread will always work~~
* ~~batch migrate data objects~~ 
* ~~Not work when more than one loops in parallel~~
* ~~Need to merge into one loop for a iteration~~

phase 2: fully automatic 
* carry over all inflight states 
* Disable by default (0 threshold)
* Allow wf start to enable manually,  with a signal channel to update later 

Notes to check:
* [ ] need to fix runID concept 
  * return originalRunId/firstRunId? may rename to be more clear
* search attributes continueAsNew: 
  * [x] carried over automatically by Temporal workflow
  * [x] carried over automatically by Temporal workflow
* [x] `manualDeferFn` can now be removed because no longer needed (it won't be correct if the state need to be carried over)
* [x] need comment for `awaitError := provider.Await(ctx, func() bool {` , why `len(currentStates) > 0` ???
  * `len(currentStates) == 0` won't work because it will meet right after the stateToExecute are poped from queue
  * It is inside the outer loop of `for len(currentStates) > 0 {` , when it reach `len(currentStates) > 0`, the outer loop will then go to the next iteration
  * we need some condition to wait here because all the stateToExecute are running in different thread
  * alternatively, using `totalPendingCount == 0` will work, but not as efficient as this one -- waiting for all current pending to complete will take much longer time
* check if querying previous run that continuedAsNew to current run is allowed in Cadence/Temporal
  * [x] works in Temporal 
  * [x] works in Cadence

**Comment by longquanzheng** [comment]
Step 1: provide query handler to expose all carry over data so that next run can get them over activity (avoid exploding the wf input limit):

* GetDataObjects (already implemented)
* GetSearchAttributes
* GetInterStateChannelPendingData
* GetSignalChannelPendingSignals 
  * opt1: hack through history
  * op2: use WorkflowInboundInterceptor??, but it doesn't exist in Cadence

**Comment by longquanzheng** [comment]
Future: add a api on top of query handler to cache data for activity pagination in next run

**Comment by longquanzheng** [comment]
Note: continue as new query handler will provide pagination

**Comment by longquanzheng** [comment]
Note: after this is done, provide infinite workflow timeout as default

**Comment by longquanzheng** [comment]
To carry over signal /interstate channel commands, need to separate the ones that have completed from incomplete, so that the received order won’t change

**Comment by longquanzheng** [comment]
pagination for query dump issues : 
1. map ordering 
2. change in the future of the dump struct 
3. too many things to go through

**Comment by longquanzheng** [comment]
Using a service api to query : 
1. Activities doesn’t need cadence temporal client 
2. Can implement cache layer in the future

**Comment by longquanzheng** [comment]
First version will do pagination in query handler because no cache implemented in api layer. 
the query with pagination is different from the one without pagination, but reuse most of the code 

the one without pagination will be useful for troubleshooting. And in the future we may move pagination to api layer then the query handler without pagination will be used for can and then remove the one with pagination

**Comment by longquanzheng** [comment]
Start wf req will provide parameter to set conAN threshold so that 
1. User can override the value 
2. Server can pass the default value from config and avoid using side effect or local activity 
3. 

Also provide a signal channel to override it later for flexibility

**Comment by longquanzheng** [comment]
https://github.com/indeedeng/iwf/pull/159

**Comment by longquanzheng** [comment]
Make sure also carry over the unapplied system signals

**Comment by longquanzheng** [comment]
https://github.com/indeedeng/iwf/pull/224

**Comment by longquanzheng** [comment]
Remove start aborted and decide aborted status types because the retry policy and timeout will be broken after continueAsNew, or will need a lot of work to fix.

**Comment by longquanzheng** [comment]
https://github.com/indeedeng/iwf/pull/227

**Comment by longquanzheng** [comment]
Todo: readme, pers test, config
